landingai/vision_pipeline.py

```diff
@@ -0,0 +1,338 @@
+            if height is None:
+                height = int(width * float(frame.image.size[1] / frame.image.size[0]))
+            if frame.image.size[0] > width or frame.image.size[1] > height:
+                frame.image = frame.image.resize((width, height))
+        return self
+
+    def save_image(self, filename_prefix: str, image_src: str = "") -> "FrameSet":
+        """Save all the images on the FrameSet to disk (as PNG)
+
+        Parameters
+        ----------
+        filename_prefix : path and name prefix for the image file
+        image_src : if empty the source image will be displayed. Otherwise the image will be selected from `other_images`
+        """
+        timestamp = datetime.now().strftime(
+            "%Y%m%d-%H%M%S"
+        )  # TODO saving faster than 1 sec will cause image overwrite
+        c = 0
+        for frame in self.frames:
+            img = frame.image if image_src == "" else frame.other_images[image_src]
+            img.save(f"{filename_prefix}_{timestamp}_{image_src}_{c}.png", format="PNG")
+            c += 1
+        return self
+
+    def show_image(self, image_src: str = "") -> "FrameSet":
+        """Open an a window and display all the images.
+        Parameters
+        ----------
+        image_src: if empty the source image will be displayed. Otherwise the image will be selected from `other_images`
+        """
+        # TODO: Should show be a end leaf?
+        for frame in self.frames:
+            if image_src == "":
+                frame.image.show()
+            else:
+                frame.other_images[image_src].show()
+
+        # # TODO: Implement image stacking when we have multiple frames (https://answers.opencv.org/question/175912/how-to-display-multiple-images-in-one-window/)
+        # """Open an OpenCV window and display all the images. This call will stop the execution until a key is pressed.
+        # Parameters
+        # ----------
+        # image_src: if empty the source image will be displayed. Otherwise the image will be selected from `other_images`
+        # """
+        # # OpenCV is full of issues when it comes to displaying windows (see https://stackoverflow.com/questions/6116564/destroywindow-does-not-close-window-on-mac-using-python-and-opencv)
+        # cv2.namedWindow("image")
+        # cv2.startWindowThread()
+        # if image_src == "":
+        #     img = cv2.cvtColor(np.asarray(self.frames[0].image), cv2.COLOR_BGR2RGB)
+        # else:
+        #     img = cv2.cvtColor(np.asarray(self.frames[0].other_images[image_src]), cv2.COLOR_BGR2RGB)
+        # cv2.imshow("Landing AI - Press any key to exit", img)
+        # cv2.waitKey(0) # close window when a key press is detected
+        # cv2.waitKey(1)
+        # cv2.destroyWindow('image')
+        # for i in range (1,5):
+        #     cv2.waitKey(1)
+
+        return self
+
+    def apply(self, function: Callable[[Frame], Frame] = lambda f: f) -> "FrameSet":
```

Comment:
any reason to set a default callable here? i.e. `lambda f: f`
it could be prone to errors and cause confusion if we don't enforce users to provide one.

Reply:
I have tried to make all methods work (and sometimes do nothing) by default.... you will see this on other functions too